src/combinatorics/simple_graph/inc_matrix.lean

```diff
@@ -0,0 +1,392 @@
+
+-- ## Incidence matrix M
+
+/-- `inc_matrix G R` is the matrix `M` such that `M i e = 1` if vertex `i` is an
+endpoint of the edge `e` in the simple graph `G`, otherwise `M i j = 0`. -/
+def inc_matrix : matrix V G.edge_set R
+| i e := if (e : sym2 V) ∈ G.incidence_set i then 1 else 0
+
+@[simp]
+lemma inc_matrix_apply {i : V} {e : G.edge_set} :
+  G.inc_matrix R i e = if (e : sym2 V) ∈ G.incidence_set i then 1 else 0 := rfl
+
+-- ### Relation between inc_matrix elements and incidence_set property
+
+lemma inc_matrix_zero {i : V} {e : G.edge_set} [char_zero R] :
+  G.inc_matrix R i e = 0 ↔ ↑e ∉ G.incidence_set i :=
+by simp only [inc_matrix, ite_eq_right_iff, ←decidable.not_imp_not,
+              forall_true_left, not_false_iff, one_ne_zero]
+
+lemma inc_matrix_one {i : V} {e : G.edge_set} [char_zero R] :
+  G.inc_matrix R i e = 1 ↔ ↑e ∈ G.incidence_set i :=
+by simp only [inc_matrix, ite_eq_left_iff, ←decidable.not_imp_not,
+              set.not_not_mem, forall_true_left, not_false_iff, zero_ne_one]
+
+-- ### One - zero properties
+
+lemma inc_matrix_not_zero {i : V} {e : G.edge_set} [char_zero R] :
+  ¬ G.inc_matrix R i e = 0 ↔ G.inc_matrix R i e = 1 :=
+by simp only [inc_matrix_zero, inc_matrix_one, set.not_not_mem]
+
+lemma inc_matrix_not_one {i : V} {e : G.edge_set} [char_zero R] :
+  ¬ G.inc_matrix R i e = 1 ↔ G.inc_matrix R i e = 0 :=
+by simp only [inc_matrix_zero, inc_matrix_one]
+
+lemma inc_matrix_zero_or_one {i : V} {e : G.edge_set} [char_zero R] :
+  G.inc_matrix R i e = 0 ∨ G.inc_matrix R i e = 1 :=
+by { rw [inc_matrix_zero, inc_matrix_one], exact (em (↑e ∈ G.incidence_set i)).symm }
+
+lemma inc_matrix_elements_mul_one {i j : V} {e : G.edge_set} [char_zero R] :
+  G.inc_matrix R i e * G.inc_matrix R j e = 1 ↔ G.inc_matrix R i e = 1 ∧ G.inc_matrix R j e = 1 :=
+begin
+  cases G.inc_matrix_zero_or_one R with H₀ H₁,
+  { rw H₀, simp only [if_t_t, mul_boole, inc_matrix_apply, zero_ne_one, false_and] },
+  { rw H₁, simp only [true_and, mul_boole, inc_matrix_apply, eq_self_iff_true] }
+end
+
+-- 1. `∑ e : E, M i e * M j e = 1`, for any two adjacent vertices `i` and `j`.
+theorem adj_sum_of_mul_inc_one {i j : V} (H_adj : G.adj i j) :
+  ∑ (e : G.edge_set), G.inc_matrix R i e * G.inc_matrix R j e = (1 : R) :=
+begin
+  simp only [inc_matrix_apply, ite_mul_ite_zero_right, sum_boole,
+             G.mem_incidence_sets_iff_eq H_adj, mul_one],
+  rw adj_iff_exists_edge_coe at H_adj,
+  rcases H_adj with ⟨e, H_e⟩,
+  simp only [←H_e, subtype.ext_iff.symm, filter_eq', filter_congr_decidable,
+             if_true, mem_univ, nat.cast_one, card_singleton]
+end
+
+-- 2. `M i e * M j e = 0`, for any two distinct non-adjacent vertices `i`, `j` and edge `e`.
+theorem inc_matrix_mul_non_adj {i j : V} {e : G.edge_set} (Hne : i ≠ j) (H_non_adj : ¬ G.adj i j)
+[char_zero R] :
+  G.inc_matrix R i e * G.inc_matrix R j e = 0 :=
+begin
+  by_cases H₁ : G.inc_matrix R i e = 0,
+  { rw [H₁, zero_mul] },
+  { rw [inc_matrix_not_zero, inc_matrix_one] at H₁,
```

Comment:
For a `by_cases`, I like having both branches fully enclosed in braces since it illuminates the structure.  It's not like medieval times with wax tablets -- we have plenty of storage space for the bytes for the indentation.